Types.cs

```diff
@@ -0,0 +1,108 @@
+using System;
+using System.Linq;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+
+namespace cogsBurger
+{
+    /*
+    [AttributeUsage(AttributeTargets.All)]
+    public class ExclusiveRangeAttribute : RangeAttribute
+    {
+
+        public ExclusiveRangeAttribute(int minimum, int maximum) : base(minimum, maximum) { }
+
+        public override bool IsValid(object value)
+        {
+            // Automatically pass if value is null or empty. RequiredAttribute should be used to assert a value is not empty.
+            if (value == null)
+            {
+                return true;
+            }
+            string s = value as string;
+            if (s != null && String.IsNullOrEmpty(s))
+            {
+                return true;
+            }
+            dynamic val = value;
+            dynamic min = Minimum;
+            dynamic max = Maximum;
+
+            if (MinExclusive && val <= min) { return false; }
+            if (MaxExclusive && val >= max) { return false; }
+            return true;
+        }
+    }
+
+
+    [AttributeUsage(AttributeTargets.All)]
+    public class StringValidationAttribute : ValidationAttribute
+    {
+        Regex Rgx;
+        Enum Enumerations;
+
+        public StringValidationAttribute(Enum enumerations, string pattern)
+        {
+            this.Rgx = new Regex(pattern);
+            this.Enumerations = enumerations;
+        }
+
+        public override bool IsValid(object value)
+        {
+            if (!System.Enum.IsDefined(Enumerations.GetType(), value)) { return false; }
+            // check regex Pattern
+            return true;
+        }
+    }
+    */
+
+
+    public struct CogsDate
+    {
+        DateTimeOffset DateTime;
+        DateTimeOffset Date;
+        Tuple<int, int> GYearMonth;
+        int GYear;
+        TimeSpan Duration;
+        Type Type;
```

Comment:
The Type would be the same for DateTime and Date, even though they are different properties. Perhaps we make an enumeration of the different available types, and set that.

public enum CogsDateType { ....... }